pkg/venafi/tpp/connector.go

```diff
@@ -1232,25 +1244,65 @@ func (c *Connector) RetrieveCertificate(req *certificate.Request) (certificates
 	}
 
 	startTime := time.Now()
+	retrieveCount := 0
 	for {
-		var retrieveResponse *certificateRetrieveResponse
-		retrieveResponse, err = c.retrieveCertificateOnce(certReq)
+		retrieveCount++
+		statusCode, _, body, err := c.request("POST", urlResourceCertificateRetrieve, certReq)
 		if err != nil {
 			return nil, fmt.Errorf("unable to retrieve: %s", err)
 		}
-		if retrieveResponse.CertificateData != "" {
-			certificates, err = newPEMCollectionFromResponse(retrieveResponse.CertificateData, req.ChainOption)
+		var resp certificateRetrieveResponse
+		err = json.Unmarshal(body, &resp)
+		if err != nil {
+			return nil, fmt.Errorf("retrieve response body is unexpected: %s, HTTP status was %d and body was: %s", err, statusCode, body)
+		}
+		switch {
+		// After many tests, we found that the only way to know if the current
+		// certificate request is stuck due to an old failed enrollment is to
+		// look for "WebSDK CertRequest" or "Fix any errors, and then click
+		// Retry." in the retrieve response. We only check this on the first
+		// retrieve because we know that these two messages always disappear
+		// after resetting.
+		case statusCode == http.StatusInternalServerError && retrieveCount == 1 &&
+			(resp.Status == "This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry." ||
+				resp.Status == "WebSDK CertRequest Module Requested Certificate"):
+			statusCode, status, body, err := c.request("POST", urlResourceCertificateReset, certificateResetRequest{
+				CertificateDN: req.PickupID,
+				Restart:       true,
+			})
+			if err != nil {
+				return nil, fmt.Errorf("while resetting certificate due to a past failed enrollment: %w", err)
+			}
+			if statusCode != http.StatusOK {
+				return nil, fmt.Errorf("while resetting certificate due to a past failed enrollment: %s, body: '%s'", status, body)
+			}
+
+			continue
+		case statusCode == http.StatusInternalServerError && resp.Status != "":
+			return nil, fmt.Errorf("enrollment failed at stage %d: %s", resp.Stage, resp.Status)
+		case statusCode == http.StatusOK:
+			certificates, err = newPEMCollectionFromResponse(resp.CertificateData, req.ChainOption)
 			if err != nil {
-				return
+				return nil, err
 			}
+
 			err = req.CheckCertificate(certificates.Certificate)
-			return
+			if err != nil {
+				return nil, err
+			}
+
+			return certificates, nil
+		case statusCode == http.StatusAccepted:
+			// Continue below.
+		default:
+			return nil, fmt.Errorf("unexpected response while retrieving: %d: %s", statusCode, body)
 		}
+
 		if req.Timeout == 0 {
-			return nil, endpoint.ErrCertificatePending{CertificateID: req.PickupID, Status: retrieveResponse.Status}
+			return nil, endpoint.ErrCertificatePending{CertificateID: req.PickupID, Status: resp.Status}
 		}
 		if time.Now().After(startTime.Add(req.Timeout)) {
```

Comment:
Can you also fix this calculation and add the 2 seconds sleep period?
```suggestion
		if time.Now().Add(2 * time.Second).After(startTime.Add(req.Timeout)) {
```

Reply:
I assumed that fixing this "bug" may break people, that's why I did not change it. But now that I think about it, I can't find any example of breakage that may happen if we fix this.

I'll do the change.